src/System.Private.CoreLib/shared/System/IO/BinaryReader.cs

```diff
@@ -298,8 +298,24 @@ public virtual int ReadInt32()
         [CLSCompliant(false)]
         public virtual uint ReadUInt32()
         {
-            FillBuffer(4);
-            return (uint)(_buffer[0] | _buffer[1] << 8 | _buffer[2] << 16 | _buffer[3] << 24);
+            if (_isMemoryStream)
+            {
+                if (_stream == null)
+                {
+                    throw Error.GetFileNotOpen();
+                }
+
+                // read directly from MemoryStream buffer
+                MemoryStream mStream = _stream as MemoryStream;
+                Debug.Assert(mStream != null, "_stream as MemoryStream != null");
+
+                return (uint)mStream.InternalReadInt32();
+            }
+            else
+            {
+                FillBuffer(4);
+                return (uint)(_buffer[0] | _buffer[1] << 8 | _buffer[2] << 16 | _buffer[3] << 24);
+            }
```

Comment:
This is creating an exact duplicate of the ReadInt32 implementation, other than adding casts to uint when returning the data.  As such, how about just moving the implementation of ReadInt32 into an InternalReadInt32 helper, and then both ReadInt32 and ReadUInt32 just call it? e.g.
```C#
public virtual int ReadInt32() => InternalReadInt32();
public virtual uint ReadUInt32() => (uint)InternalReadInt32();
private int InternalReadInt32()
{
    ... // current implementation of ReadInt32()
}
```
?

Reply:
> This also might have minor negative perf impact on the signed version. 

The signed version is not particularly fine-tuned. You should be able to make both the signed version faster and the code look better by factoring this differently and using the BinaryPrimitives helpers:

```
public virtual int ReadInt16() => BinaryPrimitives.ReadInt32LittleEndian(InternalRead(2));
public virtual int ReadUInt16() => BinaryPrimitives.ReadUInt32LittleEndian(InternalRead(2));
public virtual int ReadInt32() => BinaryPrimitives.ReadInt32LittleEndian(InternalRead(4));
public virtual int ReadUInt32() => BinaryPrimitives.ReadUInt32LittleEndian(InternalRead(4));
public virtual int ReadInt64() => BinaryPrimitives.ReadInt64LittleEndian(InternalRead(8));
public virtual int ReadUInt64() => BinaryPrimitives.ReadUInt64LittleEndian(InternalRead(8));

private ReadOnlySpan<byte> InternalFillBuffer(int length)
{
   ... the MemoryStream special casing here...
}
```

Reply:
I am probably missing something and for sure I'm not Span<> expert but I don't see how the usage of ReadOnlySpan can help me here. The perf boost is coming from calling to mStream.InternalReadInt32() and than there is no value of converting the int I already have to span just for the sake of converting it back to int by the BinaryPrimitives class.
What you might mean, but will be way bigger change is to add a function to MemoryStream to expose ReadOnlySpan<byte> and advance the internal pointer with the amount specified. This will allow improving all BinaryReader functionality. Is it what you had in mind?

Reply:
MemoryStream currently just expose int via InternalReadInt32 so it doesn't make sense to convert it to span and than covert it back to int via BinaryPrimitives. Just to make sure I am not missing some easy performance opportunity I also tried adding new MemoryStream.InternalReadSpan which just return a ReadOnlySpan and than use that in the BinaryReader but that was way slower than the current implementation. so the bottom line is that I think the current PR is as fast as it gets for reading int/unit and is aligned with the current code base.
*DISCLAIMER*: i'm not expert with Spans so maybe i'm missing something clear. Will be glad if someone think I'm missing anything.
`internal ReadOnlySpan<byte> InternalReadSpan(int count)
        {
            EnsureNotClosed();

            if (_length - _position < count)
                throw new IOException("Not enought data to read");

            var span = new ReadOnlySpan<byte>(_buffer, _position, count);

            _position += count;
            return span;
        }`

Reply:
You are right that the extra frame is going to slow things down. It is ok to use AggressiveInlining to compensate for it (e.g. when the method is small and it is called from a limited number of places).

Here is what I have in mind with Span: https://github.com/jkotas/coreclr/commit/dc04600654d1db0d43050630be89a5fcfd606f42 . The performance is very close to what's there today (a few percent slower). I think it would be acceptable to mark even the InternalRead method with AggresiveInlining. It would eliminate the extra frame and made it even faster than what is there today at the cost of some JIT code bloat.